File: Day25/Part3/mirror_code.cpp
#include <bits/stdc++.h>
using namespace std;

// The Mirror Code — flip up to k zero runs; reflection adds up to 1 on each side if outside bit is 0.

int max_ones_with_reflection(const string& s, int k){
    int n = (int)s.size(); if(n==0) return 0;
    vector<pair<int,int>> zr; // [L,R]
    for(int i=0;i<n;){
        if(s[i]=='0'){
            int j=i; while(j+1<n && s[j+1]=='0') ++j; zr.push_back({i,j}); i=j+1;
        }else ++i;
    }
    if(zr.empty()) return n;
    if(k==0){
        int best=0;
        for(int i=0;i<n;){
            if(s[i]=='1'){ int j=i; while(j+1<n && s[j+1]=='1') ++j; best=max(best, j-i+1); i=j+1; }
            else ++i;
        }
        return best;
    }
    int m = (int)zr.size();
    int ans=0; int l=0;
    for(int r=0;r<m;++r){
        while(r-l+1>k) ++l;
        int L=zr[l].first, R=zr[r].second;
        int length = R-L+1;
        if(L-1>=0 && s[L-1]=='0') ++length;
        if(R+1<n && s[R+1]=='0') ++length;
        int a=L-1; while(a>=0 && s[a]=='1'){ ++length; --a; }
        int b=R+1; while(b<n && s[b]=='1'){ ++length; ++b; }
        ans = max(ans, length);
    }
    return min(ans, n);
}

int main(){
    cout << max_ones_with_reflection(string("0101001"), 2) << "\n"; // 5
    return 0;
}

File: Day25/Part3/mirror_code.py
# The Mirror Code
# Flip up to k zero-runs to maximize consecutive 1s, with a single-step reflective extension:
# when flipping a zero-run [L,R], if s[L-1]==0 (left) add +1, if s[R+1]==0 (right) add +1.
# This heuristic matches the hint and yields the sample.
from typing import List

def max_ones_with_reflection(s: str, k: int) -> int:
    n = len(s)
    if n == 0: return 0
    # collect zero runs as intervals [L,R]
    zero_runs = []
    i = 0
    while i < n:
        if s[i] == '0':
            j = i
            while j+1 < n and s[j+1] == '0':
                j += 1
            zero_runs.append((i,j))
            i = j+1
        else:
            i += 1
    if not zero_runs:
        return n
    if k == 0:
        # return max existing ones run
        best = 0
        i = 0
        while i < n:
            if s[i] == '1':
                j = i
                while j+1 < n and s[j+1] == '1':
                    j += 1
                best = max(best, j-i+1)
                i = j+1
            else:
                i += 1
        return best
    # sliding window over zero_runs selecting up to k contiguous runs
    m = len(zero_runs)
    ans = 0
    l = 0
    for r in range(m):
        # maintain window size l..r with size <= k by moving l only if more than k
        while r - l + 1 > k:
            l += 1
        # compute span [L,R]
        L = zero_runs[l][0]
        R = zero_runs[r][1]
        span = R - L + 1
        # include ones between and inside: total covered = R-L+1 after flips, but ones between are counted automatically as they are within [L,R]
        length = span
        # reflect bonus
        if L-1 >= 0 and s[L-1] == '0':
            length += 1
        if R+1 < n and s[R+1] == '0':
            length += 1
        # but we can also extend through existing ones outside contiguous zeros: include outer ones adjacent to the span
        # extend left through consecutive '1's before L
        a = L-1
        while a >= 0 and s[a] == '1':
            length += 1
            a -= 1
        # extend right through consecutive '1's after R
        b = R+1
        while b < n and s[b] == '1':
            length += 1
            b += 1
        ans = max(ans, length)
    # If k >= number of zero runs, flipping them all makes all 1s; ensure cap at n
    return min(ans, n)

if __name__ == "__main__":
    print(max_ones_with_reflection("0101001", 2))  # 5
